### ISSUE TITLE
Add `expectError` assertion method

### ISSUE THREAD

**Issue by sindresorhus** [SamVerschueren/tsd-check#2]
Like `$ExpectError` in https://github.com/Microsoft/dtslint#write-tests

**Comment by SamVerschueren** [comment]
I'm not sure if this is possible with the way how it's currently implemented. Currently we just ask for the [`preEmitDiagnostics`](https://github.com/SamVerschueren/tsd-check/blob/master/source/lib/compiler.ts#L21) and let the TS compiler handle all the heavy lifting.

Implementing something like an `expectError()` should check if we actually have an error, if we don't, add a custom diagnostic ourselves and if it does, remove the diagnostic from the list.

So I think if we want this, it requires us to implement some more AST parsing logic.

**Comment by SamVerschueren** [comment]
I think I got it working. This is what I have right now.

```ts
function foo(a: string, b: string) {
	return a + b;
}
```

```ts
expectError(foo('foo', 'bar'));
//=> Expected an error, but found none

expectError(foo(12, 15));
```

I'm not entirely sure how `dtslint` does it, it looks like it ignores every compiler error. Currently in my implementation, I only ignore errors which are caused by wrong arguments (passing in strings instead of numbers). Not sure what to do here? @sindresorhus you probably already came across quite a lot of use cases, is this enough or should it really ignore every error?

An example is this

```ts
expectError(
	foo(12, 14
);
```

That statement has two because it misses a `)`. Should it report nothing or should it still report the missing `)`?

To be honest, for now, I would be perfectly fine with ignoring only parameter type mismatch errors.

**Comment by SamVerschueren** [comment]
Mmmm, with the current implementation, the `MergeExclusive` negative test will not work 🤔.

https://github.com/sindresorhus/type-fest/pull/6/files#diff-6758bd513790b0134f1b07f6456260dbR29

It's a valid use case, so will try to fix it.

**Comment by SamVerschueren** [comment]
Already fixed, actually it worked out of the box, only had to make `expectError()` generic :p.

**Comment by sindresorhus** [comment]
> That statement has two because it misses a ). Should it report nothing or should it still report the missing )?

It should still report missing `)`.

**Comment by SamVerschueren** [comment]
Awesome, then it works as expected. Let's consider it beta though as I'm not sure if it covers all the use cases. I'll do a release so we can start using it :)

**Comment by sindresorhus** [comment]
// @BendingBender FYI :)
